elbepack/initvmaction.py

```diff
@@ -233,7 +336,58 @@ class StopAction(InitVMAction):
     def __init__(self, node):
         InitVMAction.__init__(self, node)
 
-    def execute(self, _initvmdir, _opt, _args):
+    def _stop_qemu_vm(self, initvmdir):
+        """
+        Stop the QEMU initvm.
+
+        This method tries to stop the QEMU initvm by sending a poweroff event
+        using QEMU monitor.
+        """
+        
+        if initvmdir == None:
+            if os.path.isdir('./initvm'):
+                print('Using default initvm directory "./initvm".')
+                initvmdir = './initvm'
+            else:
+                print('No initvm found!')    
+                sys.exit(207)
+
+        # test if socat command is availalbe, and error exit if command is missing
+        try:
+            system('which socat > /dev/null 2>&1')
+        except CommandError:
+            print(f'The command "socat" is required.\nPlease install socat: "sudo apt install socat"')
+            sys.exit(208)
+        
+        # test if QEMU monitor unix-socket file exists, and error exit if not
+        if not os.path.exists(os.path.join(initvmdir, 'qemu-monitor-socket')):
+            print('No unix socket found for this vm!\nunable to shutdown this vm.')
+            if is_qemu_running():
+                # a process is bound to the soap port, but the QEMU monitor socket file is missng
+                print('There seems to be another initvm running. The soap port is in use.')
+            sys.exit(212)
+            
+        # trigger "system_powerdown" event on QEMU monitor - this causes a ACPI shutdown signal
+        if os.path.exists(os.path.join(initvmdir, 'qemu-monitor-socket')):
+            system(f'cd {initvmdir} && echo "system_powerdown" | socat - unix-connect:qemu-monitor-socket')
```

Comment:
This should use the python `socket` module instead of `socat`.

Reply:
Done for this case.
For 'attach' I stay with socket, since the libvirt version is also relying on CLI commands and no suitable Python library is avaiblable. The libtelnet was deprecated and implementing a new client for this edge-case feature makes no sense.